src/timedata.h

```diff
@@ -75,4 +76,8 @@ int64_t GetTimeOffset();
 int64_t GetAdjustedTime();
 void AddTimeData(const CNetAddr& ip, int64_t nTime);
 
+/** Functions to enable testing the contract for AddTimeData */
+int CountOffsetSamples();
+void AddTimeDataAlgorithm(const CNetAddr& ip, const int64_t nOffsetSample, std::set<CNetAddr>& knownSet, CMedianFilter<int64_t>& offsetFilter, int64_t& offset);
```

Comment:
Could move these declarations to the test file, so that production code does not have access to the declarations.

Would benefit from `EXCLUSIVE_LOCKS_REQUIRED` on `AddTimeDataAlgorithm` - which if I'm not mistaken could be applied to the body of the method rather than the declaration.